### PR TITLE
wazuh-modulesd prints child process not found error

### PR DESCRIPTION
**Pull request by nbertoldo**
|Related issue|
|---|
|#17931|

## Description

As described in #17931, it has been detected that when `wazuh-modulesd` is shut down this unexpected error may appear:

- `ossec.log`
    ```
    2023/07/07 14:11:25 wazuh-modulesd: ERROR: Child process 2450 not found.
    ```
This error can occur for two different reasons:

### Case 1
`wm_kill_children` ([wm_exec.c](https://github.com/wazuh/wazuh/blob/864298e79e706174d8fb9ffdedfa13b4769cfc08/src/wazuh_modules/wm_exec.c#L642)) kills every child process on _wazuh-modulesd_ shutdown. It zeroes all PID array positions on the same _wazuh-modulesd_ process (the `fork()`'s parent). If a command executing thread has the time to detect the process termination before _wazuh-modulesd_ definitely exits, it might try to delete the PID from the array, but it's already zeroed. This causes _wazuh-modulesd_ prints the error.
```
2023/08/17 11:54:18 wazuh-modulesd[2986023] main.c:184 at wm_cleanup(): DEBUG: Cleaning up
2023/08/17 11:54:18 wazuh-modulesd[2986023] wm_exec.c:655 at wm_kill_children(): DEBUG: Killing 2987293
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.INFO: Starting command ''.

2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd[2986023] wm_exec.c:655 at wm_kill_children(): DEBUG: Killing 2987295
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:230 at wm_command_main(): DEBUG: Command '' finished.
2023/08/17 11:54:18 wazuh-modulesd:command[2986023] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:54:18 wazuh-modulesd[2986023] wm_exec.c:635 at wm_remove_sid(): ERROR: Child process 2987293 not found.
2023/08/17 11:54:18 wazuh-modulesd[2986023] wm_exec.c:635 at wm_remove_sid(): ERROR: Child process 2987295 not found.
```
**Note:** the logs "Cleaning up" and "Killing PID" were added for testing purposes.

### Case 2
When a command creates a new child process, if the array is full, the PID is not stored. When the child process terminates, the parent tries to remove it from the array and does not find it. This causes _wazuh-modulesd_ prints the error.
```
2023/08/17 11:38:55 wazuh-modulesd:command[2970885] wm_command.c:192 at wm_command_main(): INFO: Starting command ''.
2023/08/17 11:38:55 wazuh-modulesd[2970885] wm_exec.c:617 at wm_append_sid(): ERROR: Child process pool is full. Couldn't register sid 2971012.
2023/08/17 11:38:55 wazuh-modulesd[2970885] wm_exec.c:617 at wm_append_sid(): ERROR: Child process pool is full. Couldn't register sid 2971014.
2023/08/17 11:38:55 wazuh-modulesd[2970885] wm_exec.c:635 at wm_remove_sid(): ERROR: Child process 2971012 not found.
2023/08/17 11:38:55 wazuh-modulesd[2970885] wm_exec.c:635 at wm_remove_sid(): ERROR: Child process 2971014 not found.
``` 

The changes introduced in this PR prevent the error in **Case 1**. However, if the array becomes full, the error may appear (**Case 2**).

## Tests
- Memory tests for Linux
  - [ ] Coverity
  - [x] ThreadSanitizer

### CODE REVIEW

File: src/wazuh_modules/wm_exec.c [hunk 1]
```diff
@@ -600,58 +599,50 @@ void* reader(void *args) {
...
+    pid_t * p_sid = NULL;
+    OSListNode * node_it = NULL;
 
-    w_mutex_lock(&wm_children_mutex);
-
-    for (i = 0; i < WM_POOL_SIZE; i++) {
-        sid = wm_children[i];
-
-        if (sid) {
+    OSList_foreach(node_it, wm_children_list) {
```
Comment (JorgeMarinoDev) [inline src/wazuh_modules/wm_exec.c:642]: Although we know OSList has mutexes for I/O. Here we are forking child processes to kill the pids. 
The previous code did lock a mutex in the parent process. This keeps the list safe.
But using For Each, we are not locking the list. Is this acceptable?

**Review by JorgeMarinoDev — CHANGES_REQUESTED** [review]
LGTM AT a general level. 
Race Condition question. Analyze at functional level.
We need to lock the list before killing the child pids.

File: src/wazuh_modules/wm_exec.c [hunk 2]
```diff
@@ -32,18 +30,31 @@ typedef struct ThreadInfo {
...
+// 
+static void wm_children_node_clean(pid_t *p_sid) {
+    os_free(p_sid);
+}
+
 // Initialize children pool
 
 void wm_children_pool_init() {
-    w_mutex_init(&wm_children_mutex, NULL);
+    wm_children_list = OSList_Create();
```
Comment (JorgeMarinoDev) [inline src/wazuh_modules/wm_exec.c:43]: can this fail ?

File: src/wazuh_modules/wm_exec.c [hunk 3]
```diff
@@ -32,18 +30,31 @@ typedef struct ThreadInfo {
...
+    OSList_SetFreeDataPointer(wm_children_list, (void (*)(void *))wm_children_node_clean);
+}
+
+// Destroy children pool
+
+void wm_children_pool_destroy() {
+    OSList_Destroy(wm_children_list);
```
Comment (JorgeMarinoDev) [inline src/wazuh_modules/wm_exec.c:50]: do we add wm_children_list = NULL afterward?
do we need this function at all ?

File: src/wazuh_modules/wm_exec.c [hunk 4]
```diff
@@ -206,57 +217,47 @@ DWORD WINAPI Reader(LPVOID args) {
...
-    int i;
-
-    w_mutex_lock(&wm_children_mutex);
+    HANDLE * p_hProcess = NULL;
+    OSListNode * node_it = NULL;
 
-    for (i = 0; i < WM_POOL_SIZE && wm_children[i]; i++)  {
-        TerminateProcess(wm_children[i], 127);
+    OSList_foreach(node_it, wm_children_list) {
+        p_hProcess = (HANDLE *)node_it->data;
```
Comment (JorgeMarinoDev) [inline src/wazuh_modules/wm_exec.c:258]: can we add a simple check  if (node_it->data) before?

File: src/wazuh_modules/wm_exec.c [hunk 5]
```diff
@@ -690,22 +680,22 @@ void wm_kill_children() {
...
                 }
             } else {
                 // Kill immediately
                 kill(-sid, SIGKILL);
             }
         }
     }
-
-    w_mutex_unlock(&wm_children_mutex);
+    wm_children_pool_destroy();
```
Comment (JorgeMarinoDev) [inline src/wazuh_modules/wm_exec.c:698]: do we need this at all if we can not use the same over again? adding extra checks for this is a good idea if we need to.

File: src/wazuh_modules/wm_exec.c (hunk 4 above)
Comment (LucioDonda) [reply src/wazuh_modules/wm_exec.c:263]: will do

File: src/wazuh_modules/wm_exec.c (hunk 1 above)
Comment (LucioDonda) [reply src/wazuh_modules/wm_exec.c:652]: Right, will fix!

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

**Comment by LucioDonda** [comment]
👍🏽 LGTM but waiting for a needed rebase

**Review by nmkoremblum — CHANGES_REQUESTED** [review]
I cannot build the source dependencies, should we rebase the branch?

**Comment by nmkoremblum** [comment]
Following the steps mentioned by @vikman90 [here](https://github.com/wazuh/wazuh/issues/17931#issuecomment-1671636738), I added the hundreds of commands and then killed `wazuh-modulesd`and restarted it with the command: `for i in 1..20; do /var/ossec/bin/wazuh-modulesd -fd & pid=$!; sleep 1; kill $pid; done` and I am getting the following errors:
```bash
root@30-u22-manager:~# tail -f /var/ossec/logs/ossec.log | grep "ERROR"
2024/04/25 13:24:39 wazuh-modulesd: ERROR: Child process 473819 not found.
2024/04/25 13:25:21 wazuh-modulesd:vulnerability-scanner[474134] vulnerabilityScannerFacade.cpp:594 at start(): ERROR: VulnerabilityScannerFacade::start: Couldn't find column family: 'vendor_map'
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476701 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476702 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476697 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476698 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476708 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476713 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476704 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476706 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476715 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476719 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476723 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476711 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476717 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476844 could not be registered.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:670 at wm_remove_sid(): ERROR: Child process 476721 not found.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476845 could not be registered.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476847 could not be registered.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476849 could not be registered.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476851 could not be registered.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476853 could not be registered.
2024/04/25 13:25:26 wazuh-modulesd[474134] wm_exec.c:646 at wm_append_sid(): ERROR: Child process sid 476856 could not be registered.
``

File: src/wazuh_modules/wm_exec.c [hunk 6]
```diff
@@ -616,58 +632,61 @@ void* reader(void *args) {
...
-        if (!wm_children[i]) {
-            wm_children[i] = sid;
-            break;
+    os_calloc(1, sizeof(pid_t), p_sid);
+    *p_sid = sid;
+
+    if (wm_children_list) {
+        if (OSList_AddData(wm_children_list, (void *)p_sid)) {
+            return;
         }
```
Comment (vikman90) [inline src/wazuh_modules/wm_exec.c:644]: I miss a mutex lock here.

```suggestion
        w_mutex_lock(&wm_children_mutex);
        void * retval = OSList_AddData(wm_children_list, (void *)p_sid);
        w_mutex_unlock(&wm_children_mutex);

        if (retval != NULL) {
            return;
        }
```

File: src/wazuh_modules/wm_exec.c [hunk 7]
```diff
@@ -222,57 +237,59 @@ DWORD WINAPI Reader(LPVOID args) {
...
+    os_calloc(1, sizeof(HANDLE), p_hProcess);
+    *p_hProcess = hProcess;
 
-    for (i = 0; i < WM_POOL_SIZE; i++) {
-        if (!wm_children[i]) {
-            wm_children[i] = hProcess;
-            break;
+    if (wm_children_list) {
+        if (OSList_AddData(wm_children_list, (void *)p_hProcess)) {
+            return;
```
Comment (vikman90) [inline src/wazuh_modules/wm_exec.c:247]: Possible missing lock.

```suggestion
        w_mutex_lock(&wm_children_mutex);
        void * retval = OSList_AddData(wm_children_list, (void *)p_hProcess);

        w_mutex_unlock(&wm_children_mutex);
        if (retval != NULL) {
            return;
```

**Review by vikman90 — CHANGES_REQUESTED** [review]
@nmkoremblum The weird behavior you're reporting might be related to these missing guards at critical sections.

Please consider adding them, and also running ThreadSanitizer.

**Review by nmkoremblum — DISMISSED** [review]
LGTM

**Review by nmkoremblum — APPROVED** [review]
LGTM :rocket:

File: src/unit_tests/wazuh_modules/exec/test_wm_exec.c [hunk 8]
```diff
@@ -66,10 +105,355 @@ static void test_wm_exec_not_accented_command(void ** state) {
...
+    will_return(__wrap_OSList_GetFirstNode, node);
+    expect_function_call(__wrap_OSList_DeleteThisNode);
+
+    wm_remove_sid(sid);
+
+    os_free(node);
+}
+
+static void test_wm_kill_children_fork_failed(void ** state) {
+    pid_t sid = 1234;
```
Comment (LucioDonda) [inline src/unit_tests/wazuh_modules/exec/test_wm_exec.c:179]: we could unify this `pid_t sid = 1234;` as a global variable, right?

**Review by LucioDonda — APPROVED** [review]
Simple comment on tests, but it LGTM!

**Review by cborla — APPROVED** [review]
LGTM

**Review by vikman90 — APPROVED** [review]

**Pull request merged**